content/advent-2015/reducing-boilerplate-with-go-generate.md

```diff
@@ -0,0 +1,294 @@
+our source code and `go generate` will do the work for us.
+
+We just need to add a comment with the following format:
+
+```
+//go:generate shell command
+```
+
+Once you have that, `go generate` will automatically call `command`
+whenever it's run.
+
+There's two points that are important to remember:
+
+* `go generate` is meant to be run by the developer authoring the
+  program or package. It's never called automatically by `go get`
+* You need to have all the tools invoked by `go generate` already
+  installed and setup in your system. Make sure you document what tools
+  you are going to use and where those tools can be downloaded.
+
+Also, if your code generation tool is within the same repository, I
+would recommend calling `go run` from `go:generate`. That way, you can run
+`generate` without building and installing the tool manually each time
+you change the tool.
+
+# How do you start building your own tools?
+
+The stdlib packages to parse and generate code are great, but their
+documentation is huge and it can be quite daunting to make a sense of
+how to use the packages just from the docs.
+
+The best thing I did when I got into code generation was to learn about
+some of the existing tools. It serves three purposes:
+
+1. You'll get some inspiration about the kind of tools you can build.
+2. You'll have the chance to learn from the tools' source code.
+3. You can find some of these tools really useful by themselves.
+
+# Projects to learn from
+
+## Generating stubs to implement an interface
+
+Have you ever found yourself copying and pasting the list of methods
+defined in an interface you've got to implement?
+
+You can use [`impl`][impl] to generate the stubs automatically. It'll
+use the packages in the stdlib to look for the interface and output
+methods you must implement.
+
+```bash
+$ impl 'f *File' io.ReadWriteCloser
+func (f *File) Read(p []byte) (n int, err error) {
+    panic("not implemented")
+}
+
+func (f *File) Write(p []byte) (n int, err error) {
+    panic("not implemented")
+}
+
+func (f *File) Close() error {
+    panic("not implemented")
+}
+```
+
+## Generating mocks automatically with mockery
+
+[testify][testify] has a nice [mock][testify-mock] package to
+
+Here's a very simplified example about how to mock a service:
+
+```go
+package main
+
+import (
+  "testing"
+
+  "github.com/stretchr/testify/mock"
+)
+
+type downcaser interface {
+  Downcase(string) (string, error)
+}
+
+func TestMock(t *testing.T) {
+  m := &mockDowncaser{}
+  m.On("Downcase", "FOO").Return("foo", nil)
+  m.Downcase("FOO")
+  m.AssertNumberOfCalls(t, "Downcase", 1)
+}
+```
+
+The implementation of the mock is pretty straightforward:
+
+```go
+type mockDowncaser struct {
+  mock.Mock
+}
+
+func (m *mockDowncaser) Downcase(a0 string) (string, error) {
+  ret := m.Called(a0)
+  return ret.Get(0).(string), ret.Error(1)
+}
+```
+
+However, as we can see from the implementation, it's so straightforward
```

Comment:
Actually??????, as we can see from the implementation, it's so straightforward


There's no opposition with the previous sentence